<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:right;border-radius:5px 5px;">
    <i>Aprender una técnica no es un fin en si mismo, simplemente indica por donde hay que empezar</i><br><b>Proverbio Japonés</b></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:right;border-radius:5px 5px;">
<i>Material curado y organizado por el Prof. Dr. Hernán D. Merlino</i><br> 
<i>Exploratory Data Analysis Versión 1.0 - 2022 Q3</i></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:25px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>Exploratory Data Analysis</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
El Análisis Exploratorio de Datos (<b><i>EDA</i></b> - Exploratory Data Analysis) es la etapa en el proceso de <b><i>ciencia de datos</i></b> que evalua los datos utilizando técnicas visuales, aunque se suelen usar técnicas no visuales  en menor grado.</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Importancia</u></strong>:
<ul style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<li><i>Alta</i></li>
</ul></p>

<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Otros Nombre Utilizados como Sinonimos:</u></strong><br> 
<ul style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<li><i>Data Exploration</i></li>
<li><i>Data Visualization</i></li>
</ul></p>
    
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Secuencia en el proceso de Ciencia de Datos:</u></strong><br> 
<ul style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<li>Luego del <b><i>Data Wrangling</i></b></li>
<li>Antes del <b><i>Data Cleaning</i></b></li> 
</ul></p>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Nivel de Conocimiento Necesario:</u></strong><br>
Cuando mas experticia se obtenga en esta estapa sera mejor el resultado obtenido en el proceso de <b><i>ciencia de datos</i></b>.<BR> La manera práctica de obtener experticia es buscando dataset en Internet y probar.<BR> La idea es empezar a reconocer situaciones habituales en los conjuntos datos.</p>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura obligatoria:</u></strong><br> 
<a href="https://towardsdatascience.com/exploratory-data-analysis-in-python-c9a77dfa39ce" target=\"_blank\">Exploratory data analysis in Python</a><br>
<a href="https://medium.com/data-folks-indonesia/10-things-to-do-when-conducting-your-exploratory-data-analysis-eda-7e3b2dfbf812" target=\"_blank\">10 Things to do when conducting your Exploratory Data Analysis (EDA)</a>
</p><BR>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
Sobre la última referencia, se extrae una lista de actividades comunes:<br> 
<ul style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<li>Cambiar tipo de datos</li>
<li>Cantidad de filas y columnas</li> 
<li>Detectar anomalías</li>    
<li>Dividir columnas</li>
<li>Evaluar distribuciones</li>
<li>Evaluar correlaciones</li>    
<li>Sumarizar estadisticas</li> 
<li>Validar valores nulos</li> 
<li>Verificar duplicados</li> 
</ul>
</p>  
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario sobre actividades comunes:</u></strong><br> 
No es una lista exaustiva solo se da como ejemplo, el orden es alfabético se implementa segun la necesidad.</p><BR>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia Bibliográfica:</u></strong><br> 
<a href="https://www.packtpub.com/product/hands-on-exploratory-data-analysis-with-python/9781789537253?_ga=2.208624657.2096903426.1661512488-836029480.1661512488
" target=\"_blank\">Hands-On Exploratory Data Analysis with Python - ISBN 9781789537253</a><br>
<a href="https://www.packtpub.com/product/hands-on-data-analysis-with-pandas/9781800563452?_ga=2.207919249.2096903426.1661512488-836029480.1661512488
" target=\"_blank\">Hands-On Data Analysis with Pandas, Second Edition - ISBN 9781800563452</a>
</p>    


<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Bibliotecas utilizadas en el notebook</strong><br> 
<ol style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<li>pandas</li> 
<li>numpy</li> 
<li>matplotlib</li>
<li>joblib</li>
</ol>
</p> 
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Cargar librerias Pandas, Numpy y Warnings</strong></p> 

In [2]:
import pandas as pd
#Configuración de parametros de Pandas para mejor visualización
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)
pd.set_option('display.precision', 3)
pd.set_option('plotting.backend', 'matplotlib') 
pd.options.mode.chained_assignment = None

# En produccion eliminar, se incluye por informacion
print(f'Pandas Version: {pd.__version__}') 

Pandas Version: 1.5.3


<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br> 
Se esta importando la libreria <b><i>pandas</i></b> y se le esta dando un nombre reducido para invocarla, solo a modo de comodidad.<br>
Dando una definición informal, se puede decir que <b><i>pandas</i></b> es como tener una planilla de cálculo en un programa, se puede trabajar de manera similar pensando en filas y columnas y aplicando acciones sobre estas.<br></p>

<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Nivel de Conocimiento Necesario:</u></strong><br>
La libreria <b>Pandas</b> es central en <b><i>ingeniería de datos</i></b> y <b><i>ciencia de datos</i></b>, su conocimiento es básico, cuanto más se sepa mejor, nunca es suficiente y periodicamente se debe releer la referencia.</p>

<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura obligatoria:</u></strong><br> 
<a href="https://pandas.pydata.org/pandas-docs/stable/" target=\"_blank\">Pandas Documentación</a></p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario sobre Parametros de Visualización:</u></strong><br> 
Existen gran cantidad de configuraciones posibles, no solamente para la visualización<br>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Nivel de Conocimiento Necesario:</u></strong><br>
<b>Set_option & Options</b> son funciones de nivel de experticía de media a avanzada en la <b><i>ingeniería de datos</i></b>, su conocimiento se da naturalmente con el uso de la librería, se recomeinda leer la referencia para tener una idea de las posibilidades.</p>    
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura recomendada:</u></strong><br>     
<a href="https://pandas.pydata.org/docs/user_guide/options.html" target=\"_blank\">Pandas Options and settings</a></p>

In [3]:
import numpy as np

np.set_printoptions(suppress=True)

# En produccion eliminar, se incluye por informacion
print(f'Numpy version: {np.__version__}')

Numpy version: 1.24.4


<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br> 
Se esta importando la libreria <b><i>numpy</i></b> y se le esta dando un nombre reducido para invocarla.<br>
Numpy facilita el trabajo relacionado a funciones matemáticas y manejo de conjuntos de datos.</p>
<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Nivel de Conocimiento Necesario:</u></strong><br>
La libreria <b><i>Numpy</i></b> es muy utilizada en <b><i>ingeniería de datos</i></b> y <b><i>ciencia de datos</i></b>, su conocimiento es básico, pero a diferencia de <b><i>pandas</i></b>, se puede inciar el camino en la <b><i>ciencia de datos</i></b> sin un conocimiento profundo de la misma, la experticia se adquiere en la medida de su uso.</p>
<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura recomendada:</u></strong><br> 
<a href="https://numpy.org/doc/" target=\"_blank\">Numpy Documentación</a></p>

In [4]:
# En produccion eliminar, se incluye por informacion
import matplotlib
print(f'Matplotlib version: {matplotlib.__version__}')

import matplotlib.pyplot as plt
%matplotlib inline
plt.rcParams["figure.figsize"] = (15,7)

Matplotlib version: 3.7.3


<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br> 
Se esta importando la libreria <b><i>matplotlib</i></b> y se le esta dando un nombre reducido para invocarla.<br>
Matoplotlib facilita el trabajo de visualización de datos, es por defecto la librería que utiliza pandas.<br>
    El comando: <b><i>%matplotlib inline</i></b> es para que se vea en el notebook los gráficos.
</p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Nivel de Conocimiento Necesario:</u></strong><br>
La libreria <b><i>matplotlib</i></b> es muy utilizada en <b><i>ingeniería de datos</i></b> y <b><i>ciencia de datos</i></b>, su conocimiento es básico, pero a diferencia de <b><i>pandas</i></b>, se puede inciar el camino en la <b><i>ciencia de datos</i></b> sin un conocimiento profundo de la misma, la experticia se adquiere en la medida de su uso.</p>
<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura recomendada:</u></strong><br> 
<a href="https://matplotlib.org/stable/index.html/" target=\"_blank\">Matplotlib Documentación</a></p>

In [5]:
import warnings
warnings.filterwarnings("ignore")

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br> 
Esta librería se incluye solo para una mejor visualización del notebook.<BR>
En producción se deben evaluar los potenciales warnings.
</p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>____________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Lectura mediante Pandas del archivo de datos</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px;">
<strong>Bajar el dataset de Kaggle:</strong>
<a href="https://www.kaggle.com/code/fabiendaniel/customer-segmentation/data" target=\"_blank\">Customer Segmentation Data</a>
</p>    

In [6]:
df = pd.read_csv('./data/01AEDAdata.csv', encoding='unicode_escape')  

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br> 
Existen gran cantidad de configuraciones posibles, en este caso se debió agregar la opción <b><i>encoding</i></b> , pues por la configuración del equipo donde se desarrollo este programa se producia un error en la lectura del archivo.</p>
<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Nivel de Conocimiento Necesario:</u></strong><br>
<b>Read_csv</b> es una función básica para trabajar en la <b><i>ingeniería de datos</i></b>, su conocimiento es la medida de la expertícia adquirida.</p>
<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura obligatoria:</u></strong><br> 
<a href="https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html" target=\"_blank\">Pandas Read CSV</a></p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Visualizar el Dataframe:</u></strong><br> 
Es aconsejable evaluar el dataframe tanto sea al comienzo al final y de forma aleatoria, utilizar los comandos: 
<b><i>head-sample-tail</i></b></p>
</p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Visualizar las primeras 5 filas  del dataset</strong></p> 

In [7]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Visualizar de manera aleatoria 5 filas del dataset</strong></p> 

In [8]:
df.sample(5) 

InvoiceNo StockCode                        Description  Quantity  \
203304    554514     22090            PAPER BUNTING RETROSPOT         3   
167824    551005     22637              PIGGY BANK RETROSPOT          1   
481916   C577388     22423           REGENCY CAKESTAND 3 TIER        -1   
538098    581397     21121  SET/10 RED POLKADOT PARTY CANDLES        24   
65851     541696     10125            MINI FUNKY DESIGN TAPES         3   

             InvoiceDate  UnitPrice  CustomerID         Country  
203304   5/24/2011 15:58       2.95         NaN  United Kingdom  
167824    4/26/2011 9:48       5.79         NaN  United Kingdom  
481916  11/18/2011 16:58      12.75     14911.0            EIRE  
538098   12/8/2011 13:25       1.25     17668.0  United Kingdom  
65851    1/20/2011 18:08       0.79         NaN  United Kingdom

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Visualizar las ulitmas 5 filas del dataset</strong></p> 

In [9]:
df.tail()

InvoiceNo StockCode                      Description  Quantity  \
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID Country  
541904  12/9/2011 12:50       0.85     12680.0  France  
541905  12/9/2011 12:50       2.10     12680.0  France  
541906  12/9/2011 12:50       4.15     12680.0  France  
541907  12/9/2011 12:50       4.15     12680.0  France  
541908  12/9/2011 12:50       4.95     12680.0  France

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Objetivo del comando</strong><br>
Como práctica habitual se ejecuta <b><i>df.head()</i></b> , como primer comando inmediatamente despues de la carga del dataset para validar si en el proceso de carga (<b><i>pd.read_csv()</i></b>) no se produjo ningun error.</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:20px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>Comienzo del proceso de exploración de datos</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Pandas Info</strong>: evalúa los tipos de columnas (Feature/Característica/Rasgo) del dataframe</p> 

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura recomendada:</u></strong><br> 
<a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html" target=\"_blank\">Pandas Info</a></p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Comando astype()</strong><br>
Permite cambiar de tipo de datos a otro deseado.</p> 

In [11]:
df[['Quantity']] = df[['Quantity']].astype(int)

In [12]:
df[['InvoiceDate']] = df[['InvoiceDate']].astype('datetime64[ns]')

In [13]:
df['CustomerID'] = df['CustomerID'].fillna(0)               # como hay valores nulos primero se los llena con cero
df[['CustomerID']] = df[['CustomerID']].astype(int)

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
En muchos métodos que son invocados en <b>Pandas</b> se verá que hay corchetes dobles eso indica que a ese método se le puede pasar una lista de columnas para que se ejecute el método.<BR>
En este caso aunque se ha pasado una sola columna se mantiene ese concepto de conjunto de columnas, solo a fines de coherencia en el código.
</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura recomendada:</u></strong><br> 
<a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html" target=\"_blank\">Pandas Astype</a></p>

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int32         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   541909 non-null  int32         
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(1), int32(2), object(4)
memory usage: 28.9+ MB


<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Comando info()</strong><br>
Si se observa la última línea del comando:<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;memory usage: <strong>33.1+ MB</strong> : para la primera invocación<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;memory usage: <strong>28.9+ MB</strong> : para la segunda invocación<br>   
Es la reducción de consumo de memoria que se produjo al ajustar los tipos, en este ejemplo es insignificante, en grandes dataframes puede ser hacer la diferencia entre poder entrenar un modelo.</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Se sigue con el análisis exploratorio</strong><br>
En esta etapa no se sigue ningún modelo específico, se debe ser curioso para poder entender de la mejor manera los datos.</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Consultar valores únicos por columna</strong></p> 

In [15]:
np.sort(df['InvoiceNo'].unique())

array(['536365', '536366', '536367', ..., 'C581499', 'C581568', 'C581569'],
      dtype=object)

In [16]:
np.sort(df['StockCode'].unique())

array(['10002', '10080', '10120', ..., 'gift_0001_40', 'gift_0001_50',
       'm'], dtype=object)

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Se observa algo llamativo, la letra m, entonces me enfoco en ello</strong></p> 

In [17]:
df[df['StockCode'].str.contains('m')]

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
40383    539736         m      Manual         1 2010-12-21 15:18:00   

       UnitPrice  CustomerID         Country  
40383       2.55           0  United Kingdom

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>El Stockcode m, esta realcionado al ainvoceNo 539736, entonces me enfoco en ello</strong></p> 

In [18]:
df[df['InvoiceNo']== '539736']

InvoiceNo StockCode                          Description  Quantity  \
40307    539736     21035      SET/2 RED RETROSPOT TEA TOWELS          1   
40308    539736     21041      RED RETROSPOT OVEN GLOVE DOUBLE         2   
40309    539736     21116                         OWL DOORSTOP         1   
40310    539736     21154            RED RETROSPOT OVEN GLOVE          4   
40311    539736     21166           COOK WITH WINE METAL SIGN          1   
40312    539736     21169      YOU'RE CONFUSING ME METAL SIGN          1   
40313    539736     21172                    PARTY METAL SIGN          1   
40314    539736     21181         PLEASE ONE PERSON METAL SIGN         6   
40315    539736     21207       SKULL AND CROSSBONES  GARLAND          1   
40316    539736     21218               RED SPOTTY BISCUIT TIN         1   
40317    539736     21259          VICTORIAN SEWING BOX SMALL          1   
40318    539736     21411          GINGHAM HEART  DOORSTOP RED         1   
40319    539736     21484          CHICK GREY HOT WATER BOTTLE         5   
40320    539736     21494       ROTATING LEAVES T-LIGHT HOLDER         5   
40321    539736     21700          BIG DOUGHNUT FRIDGE MAGNETS         2   
40322    539736     21832                 CHOCOLATE CALCULATOR         2   
40323    539736     21876                        POTTERING MUG         3   
40324    539736     21890    S/6 WOODEN SKITTLES IN COTTON BAG         1   
40325    539736     21894          POTTING SHED SEED ENVELOPES         1   
40326    539736     21935                   SUKI  SHOULDER BAG         7   
40327    539736     21948             SET OF 6 CAKE CHOPSTICKS         1   
40328    539736     22166  WALL MIRROR RECTANGLE DIAMANTE PINK         1   
40329    539736     22179              SET 10 LIGHTS NIGHT OWL         1   
40330    539736     22191               IVORY DINER WALL CLOCK         1   
40331    539736     22197                 SMALL POPCORN HOLDER         8   
40332    539736     22207                FRYING PAN UNION FLAG         1   
40333    539736     22297            HEART IVORY TRELLIS SMALL         1   
40334    539736     22309                 TEA COSY RED  STRIPE         1   
40335    539736     22352    LUNCH BOX WITH CUTLERY RETROSPOT          1   
40336    539736     22358              KINGS CHOICE TEA CADDY          1   
40337    539736     22366                     DOORMAT AIRMAIL          2   
40338    539736     22383              LUNCH BAG SUKI  DESIGN          3   
40339    539736     22431           WATERING CAN BLUE ELEPHANT         1   
40340    539736     22457      NATURAL SLATE HEART CHALKBOARD          2   
40341    539736     22627                  MINT KITCHEN SCALES         1   
40342    539736     22629                  SPACEBOY LUNCH BOX          2   
40343    539736     22659              LUNCH BOX I LOVE LONDON         3   
40344    539736     22666      RECIPE BOX PANTRY YELLOW DESIGN         3   
40345    539736     22672      FRENCH BATHROOM SIGN BLUE METAL         1   
40346    539736     22697      GREEN REGENCY TEACUP AND SAUCER         1   
40347    539736     22699     ROSES REGENCY TEACUP AND SAUCER          1   
40348    539736     22784                LANTERN CREAM GAZEBO          2   
40349    539736     22837           HOT WATER BOTTLE BABUSHKA          1   
40350    539736     22891                 TEA FOR ONE POLKADOT         1   
40351    539736     22896                PEG BAG APPLES DESIGN         2   
40352    539736     22898        CHILDRENS APRON APPLES DESIGN         1   
40353    539736     22960             JAM MAKING SET WITH JARS         1   
40354    539736     22961               JAM MAKING SET PRINTED         2   
40355    539736     22962                JAM JAR WITH PINK LID         5   
40356    539736     22963               JAM JAR WITH GREEN LID         2   
40357    539736    35004B          SET OF 3 BLACK FLYING DUCKS         1   
40358    539736     37370           RETRO COFFEE MUGS 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>A simple vista no se observa nada fuera de lo normal, puede ser un cargo específico sin mas.</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Se observan algunos StockCode que finalizan con letras, entonces me enfoco en ello.</strong></p> 

In [19]:
df[df['StockCode'].str.contains('85160')]

InvoiceNo StockCode                   Description  Quantity  \
40382     539736    85160a  WHITE BIRD GARDEN DESIGN MUG         1   
54916     540977    85160B  BLACK BIRD GARDEN DESIGN MUG         1   
71504     542137    85160A  WHITE BIRD GARDEN DESIGN MUG         6   
75405     542595    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
109346    545600    85160A                           NaN       -21   
123636    546910    85160B  BLACK BIRD GARDEN DESIGN MUG         2   
124882    546995    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
143864    548714    85160B  BLACK BIRD GARDEN DESIGN MUG         4   
150334    549425    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
153674    549787    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
224084    556503    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
252802    559138    85160B  BLACK BIRD GARDEN DESIGN MUG        12   
255397    559342    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
257690    559542    85160B  BLACK BIRD GARDEN DESIGN MUG        12   
260232    559804    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
313759    564536    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
347575    567310    85160B  BLACK BIRD GARDEN DESIGN MUG         3   
369947   C569112    85160B  BLACK BIRD GARDEN DESIGN MUG        -2   
458379    575837    85160B  BLACK BIRD GARDEN DESIGN MUG         6   
458417    575837    85160B  BLACK BIRD GARDEN DESIGN MUG        12   
464136    576078    85160B  BLACK BIRD GARDEN DESIGN MUG         6   

               InvoiceDate  UnitPrice  CustomerID         Country  
40382  2010-12-21 15:18:00      16.13           0  United Kingdom  
54916  2011-01-12 15:01:00      16.13           0  United Kingdom  
71504  2011-01-25 17:10:00       1.25           0  United Kingdom  
75405  2011-01-30 10:19:00       1.25       14040  United Kingdom  
109346 2011-03-04 11:45:00       0.00           0  United Kingdom  
123636 2011-03-17 19:46:00       0.39       13969  United Kingdom  
124882 2011-03-18 13:25:00       0.39       14041  United Kingdom  
143864 2011-04-03 15:07:00       0.39       17337  United Kingdom  
150334 2011-04-08 11:54:00       0.39       14030  United Kingdom  
153674 2011-04-12 11:58:00       0.39       13268  United Kingdom  
224084 2011-06-13 08:44:00       0.39       13517  United Kingdom  
252802 2011-07-06 13:35:00       0.39       16919  United Kingdom  
255397 2011-07-07 16:59:00       0.39       14493  United Kingdom  
257690 2011-07-10 13:32:00       0.39       17126  United Kingdom  
260232 2011-07-12 14:17:00       0.39       17405  United Kingdom  
313759 2011-08-25 15:12:00       0.39       13159  United Kingdom  
347575 2011-09-19 13:45:00       0.39       15504  United Kingdom  
369947 2011-09-30 12:47:00       0.39       17405  United Kingdom  
458379 2011-11-11 11:37:00       0.39       12748  United Kingdom  
458417 2011-11-11 11:37:00       0.39       12748  United Kingdom  
464136 2011-11-13 16:22:00       0.39       14432  United Kingdom

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Con lo explorado al momento, see entiende que no see puede convertir el StockCode a numérico, pues hay valores combinados.</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Consulto el valor -21 que se detectó.</strong></p> 

In [20]:
df[df['InvoiceNo'].str.contains('545600')]

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
109346    545600    85160A         NaN       -21 2011-03-04 11:45:00   

        UnitPrice  CustomerID         Country  
109346        0.0           0  United Kingdom

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Se evalua una invoceNo que empeiza con la letra C.</strong></p> 

In [21]:
df[df['InvoiceNo'].str.contains('C569112')]

InvoiceNo StockCode                          Description  Quantity  \
369942   C569112     23425          STORAGE TIN HOME SWEET HOME        -2   
369943   C569112     22720    SET OF 3 CAKE TINS PANTRY DESIGN         -1   
369944   C569112     22364         GLASS JAR DIGESTIVE BISCUITS        -1   
369945   C569112     22362         GLASS JAR PEACOCK BATH SALTS        -3   
369946   C569112     22357             KINGS CHOICE BISCUIT TIN        -2   
369947   C569112    85160B         BLACK BIRD GARDEN DESIGN MUG        -2   
369948   C569112     20914  SET/5 RED RETROSPOT LID GLASS BOWLS        -1   

               InvoiceDate  UnitPrice  CustomerID         Country  
369942 2011-09-30 12:47:00       2.49       17405  United Kingdom  
369943 2011-09-30 12:47:00       4.95       17405  United Kingdom  
369944 2011-09-30 12:47:00       2.95       17405  United Kingdom  
369945 2011-09-30 12:47:00       2.95       17405  United Kingdom  
369946 2011-09-30 12:47:00       4.25       17405  United Kingdom  
369947 2011-09-30 12:47:00       0.39       17405  United Kingdom  
369948 2011-09-30 12:47:00       2.95       17405  United Kingdom

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Se podría inferir que es una nota de crédito, se debería confirmar.</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Se evalua cuantas hay.</strong></p> 

In [22]:
len(df[(df['InvoiceNo'].str.contains('C'))  & (df['Quantity']<0)])

9288

In [23]:
len(df[(df['InvoiceNo'].str.contains('C'))  & (df['Quantity']>0)])

0

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Se evaluan todos los valores negativos e intentar encontrar un patrón.</strong></p> 

In [24]:
df[(df['Quantity'] < 0)][['StockCode','Quantity', 'InvoiceDate','CustomerID']].head(10)

StockCode  Quantity         InvoiceDate  CustomerID
141         D        -1 2010-12-01 09:41:00       14527
154    35004C        -1 2010-12-01 09:49:00       15311
235     22556       -12 2010-12-01 10:24:00       17548
236     21984       -24 2010-12-01 10:24:00       17548
237     21983       -24 2010-12-01 10:24:00       17548
238     21980       -24 2010-12-01 10:24:00       17548
239     21484       -12 2010-12-01 10:24:00       17548
240     22557       -12 2010-12-01 10:24:00       17548
241     22553       -24 2010-12-01 10:24:00       17548
939     22960        -6 2010-12-01 12:38:00       17897

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Se toma un solo valor y se lo analiza en detalle.</strong></p> 

In [25]:
df[(df['StockCode']== '35004C')  & (df['CustomerID']== 15311)]

InvoiceNo StockCode                      Description  Quantity  \
154     C536383    35004C  SET OF 3 COLOURED  FLYING DUCKS        -1   
9200     537195    35004C  SET OF 3 COLOURED  FLYING DUCKS        12   
18196   C537805    35004C  SET OF 3 COLOURED  FLYING DUCKS        -1   
28630    538651    35004C  SET OF 3 COLOURED  FLYING DUCKS        12   
39381   C539640    35004C  SET OF 3 COLOURED  FLYING DUCKS        -3   
44157    540157    35004C  SET OF 3 COLOURED  FLYING DUCKS        12   
59542    541293    35004C  SET OF 3 COLOURED  FLYING DUCKS        12   
78333   C542866    35004C  SET OF 3 COLOURED  FLYING DUCKS        -2   

              InvoiceDate  UnitPrice  CustomerID         Country  
154   2010-12-01 09:49:00       4.65       15311  United Kingdom  
9200  2010-12-05 13:55:00       4.65       15311  United Kingdom  
18196 2010-12-08 13:18:00       4.65       15311  United Kingdom  
28630 2010-12-13 15:07:00       4.65       15311  United Kingdom  
39381 2010-12-20 15:27:00       4.65       15311  United Kingdom  
44157 2011-01-05 11:41:00       4.65       15311  United Kingdom  
59542 2011-01-17 13:39:00       4.65       15311  United Kingdom  
78333 2011-02-01 12:14:00       4.65       15311  United Kingdom

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Conclusión:</strong><br>
Se intento conocer el dataset y en base a la curiosisdad analizar el mismo.<br>
Esto genera que se daban contactar a los dueñso de los datos para poder hacer las preguntas que surgen del análisis.<br>
Este proceso se repite hasta que el ingendieor en datos se siente seguro con la evaluación que ha hecho.</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Comando describe()</strong><br>
Permite avanzar sobre el conocimiento del dataset, para empezar a reconocer su distribución, trabajando solamente con las columnas/features/rasgos del tipo texto/string.</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Definición de Percentiles:</strong><br>
El percentil es una medida de posición usada en estadística que indica, una vez ordenados los datos de menor a mayor, el valor de la variable por debajo del cual se encuentra un porcentaje dado de observaciones en un grupo.<br>
Ejemplo: el percentil 20º es el valor bajo el cual se encuentran el 20% de las observaciones y el 80% restante son mayores.
<br><br>
Aparecen citados en la literatura científica por primera vez por Francis Galton en 1885.<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;P25 = Q1<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;P50 = Q2 = mediana<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;P75 = Q3<br></p>

<a id = "1.0"></a>
<p style="font-size:14px;color:#16268a;font-family:'Comic Sans MS';text-align:right;border-radius:5px;">
<a href="https://es.wikipedia.org/wiki/Percentil" target=\"_blank\"><i>Referencia Wikipedia - Percentil</i></a></p>

In [26]:
df[['Quantity','UnitPrice']].describe(percentiles=[0.05, 0.25, 0.50, 0.75, 0.95])

Quantity   UnitPrice
count  541909.000  541909.000
mean        9.552       4.611
std       218.081      96.760
min    -80995.000  -11062.060
5%          1.000       0.420
25%         1.000       1.250
50%         3.000       2.080
75%        10.000       4.130
95%        29.000       9.950
max     80995.000   38970.000

In [27]:
df.describe(exclude=['object','float', 'int'], percentiles=[0.05, 0.25, 0.50, 0.75, 0.95])

InvoiceDate
count                541909
unique                23260
top     2011-10-31 14:41:00
freq                   1114
first   2010-12-01 08:26:00
last    2011-12-09 12:50:00

In [28]:
df.describe(include=['object'], exclude=['float', 'int'])

InvoiceNo StockCode                         Description         Country
count     541909    541909                              540455          541909
unique     25900      4070                                4223              38
top       573585    85123A  WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
freq        1114      2313                                2369          495478

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura recomendada:</u></strong><br> 
<a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html" target=\"_blank\">Pandas Describe</a></p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Comando unique()</strong><br>
Muestra por feature los valores únicos, sirve para mejorar la comprensión del dataset.</p>

In [29]:
%time
np.sort(df['Country'].unique())

CPU times: total: 0 ns
Wall time: 0 ns


array(['Australia', 'Austria', 'Bahrain', 'Belgium', 'Brazil', 'Canada',
       'Channel Islands', 'Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
       'European Community', 'Finland', 'France', 'Germany', 'Greece',
       'Hong Kong', 'Iceland', 'Israel', 'Italy', 'Japan', 'Lebanon',
       'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland',
       'Portugal', 'RSA', 'Saudi Arabia', 'Singapore', 'Spain', 'Sweden',
       'Switzerland', 'USA', 'United Arab Emirates', 'United Kingdom',
       'Unspecified'], dtype=object)

In [30]:
%time
np.unique(df['Country'])

CPU times: total: 0 ns
Wall time: 0 ns


array(['Australia', 'Austria', 'Bahrain', 'Belgium', 'Brazil', 'Canada',
       'Channel Islands', 'Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
       'European Community', 'Finland', 'France', 'Germany', 'Greece',
       'Hong Kong', 'Iceland', 'Israel', 'Italy', 'Japan', 'Lebanon',
       'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland',
       'Portugal', 'RSA', 'Saudi Arabia', 'Singapore', 'Spain', 'Sweden',
       'Switzerland', 'USA', 'United Arab Emirates', 'United Kingdom',
       'Unspecified'], dtype=object)

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
Aquí se ve como la librería <b>Pandas</b> utiliza a la librería <b>Numpy</b>.
</p>    

In [31]:
np.sort(df['Quantity'].unique())

array([-80995, -74215,  -9600,  -9360,  -9058,  -5368,  -4830,  -3667,
        -3167,  -3114,  -3100,  -3000,  -2880,  -2834,  -2618,  -2600,
        -2472,  -2376,  -2003,  -2000,  -1930,  -1897,  -1681,  -1671,
        -1560,  -1515,  -1512,  -1510,  -1479,  -1440,  -1430,  -1400,
        -1350,  -1300,  -1296,  -1284,  -1277,  -1206,  -1200,  -1158,
        -1131,  -1128,  -1121,  -1100,  -1092,  -1060,  -1050,  -1000,
         -990,   -975,   -967,   -960,   -939,   -905,   -900,   -864,
         -840,   -828,   -800,   -786,   -756,   -752,   -750,   -741,
         -723,   -720,   -701,   -696,   -690,   -682,   -675,   -674,
         -670,   -664,   -657,   -635,   -624,   -620,   -618,   -600,
         -590,   -576,   -553,   -550,   -541,   -530,   -524,   -504,
         -500,   -484,   -480,   -472,   -468,   -458,   -455,   -450,
         -443,   -434,   -432,   -430,   -428,   -420,   -413,   -408,
         -400,   -398,   -391,   -390,   -384,   -380,   -365,   -360,
      

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br>
El conjunto de pasos realizados anteriormente se repiten en casi todos los procesos de exploración, sirven para normalizar el dataset y para emepzar a conocerlo, es por esto que este código se debe tener al alcance de la mano siempre para utilizarlo tantas veces sea necesario.
</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:25px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>Guardar Dataset</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong>Comando to_csv()</strong><br>
Permite grabar en disco un dataset.</p> 

In [32]:
df.to_csv('./data/01AEDAdataV1.csv', encoding='latin-1', index=False, header=True, decimal='.', sep=',')

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br>
La función <b><i>to_csv</i></b> al igual que read_csv se debe conocer en detalle pues las diversas forma de guardar un archivo permiten trabajar de manera más adecuada.<br>
Es importante comprender el parametor <b>encoding</b>, este sule ser un parametro que genera problemas al momento de leer o salvar un archivo, así que se debe comprender especialmente.    
</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Referencia de lectura obligatoria:</u></strong><br> 
<a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html" target=\"_blank\">Pandas to_csv</a></p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br>
Alternativamente, se puede utilizar la librería <b>joblib</b> para generar una secuencia de trabajo (pipeline), esta librería permite grabar en disco un objeto de la memoria del programa sin transformación.<br>
La ventaja es que acelera el tiempo de grabación y posterior lectura, las desventaja es que no se puede editar, como podría ser en el cado de utilizar <b>to_csv</b>.<br>
La regla de escritorio es, en las primeras iteraciones del desarrollo utilizar <b>to_csv</b> y una vez que se avanzo y se estabilizo el dataset pasar a <b>joblib</b>.    
</p> 

In [33]:
import joblib as jb
print(f'Joblib Version: {jb.__version__}') 

Joblib Version: 1.3.2


In [34]:
jb.dump(value=df, filename='./data/01AEDAdataV1.bin')

['./data/01AEDAdataV1.bin']

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Conclusiones:</u></strong><br>
Sobre el <b><i>Análisis Exploratorio de Datos</b></i> se ha detectado:<br>
1. Importes negativos<br>
2. Descripciones de productos en blanco<br>
3. Diferencia entre el codigo de producto y las descripciones<br>
4. Correlacion entre features<br>
5. Registros duplicados<br>
<br>
Puede ser que el lector del presente documento encuentre nuevas observaciones y eso es correcto, es la esencia del <b>EDA</b>.<BR>
<br>
En esta etapa no se puede definir si estos son errores o son datos correctos, pero es necesatirio realizar un revision de los mismos y tomar una desición de que hacer. Cada una de estas desiciones debe quedar documentada.<br>
<br>
Esta es la principal razón de realizar un analisis <b>EDA</b>, se intenta de manera simplificada detectar ciertas inconcistencias que de la simple revision manual de los datos sería extremadamente difícil encontrarlas.<br>
<br>
Por otra parte, es común que este proceso haga reflexionar sobre el conjunto de datos que se tienen y se vuelva a la etapa anterior para agregar mas features, estas primeras etapas son iterativas.    
</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Proximos Pasos:</u></strong><br>
Ganar experticia en <b><i>Análisis Exploratorio de Datos</b></i>, se puede acceder a la siguiente URL para bajar otros dataset, el actual ha sido bajado de la misma página, el objetivo es duplicar este notebook y cargar el nuevo dataset e intentar encontrar potenciales inconsistencias<br>

<a href="https://www.kaggle.com/datasets" target="_blank">Kaggle Datasets</a>
<br></p>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Comentario:</u></strong><br>    
<b>Kaggle</b> es un sitio donde se desarrollan competencia de ciencia de datos, ademá se puede encontrar material para obtener mayor experticia.
</p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:left;border-radius:5px 5px;">
<strong><u>Cierre:</u></strong><br>
Existen gran cantidad de librerías de <b>AUTO EDA</b>, uno puede seguir explorando para enontrar la mas apropiada para el dataset en cuestión, pues cada una tiene distintas características.<br>
Para detectar estas librerías solo se debe buscar en Goolge: <b><i>lib python auto eda</i></b></p>

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 

<a id = "1.0"></a>
<p style="font-size:15px;color:#34656d;font-family:'Comic Sans MS';text-align:center;border-radius:5px 5px;">
<strong>_____________________________________________________________</strong></p> 